File: Python/deleteFiles.py
'''
DISCLAIMER OF WARRANTIES.
 This code is sample code created by IBM Corporation. IBM grants you a
 nonexclusive copyright license to use this sample code example. This
 sample code is not part of any standard IBM product and is provided to you
 solely for the purpose of assisting you in the development of your
 applications. This example has not been thoroughly tested under all
 conditions. IBM, therefore cannot guarantee nor may you imply reliability,
 serviceability, or function of these programs. The code is provided "AS IS",
 without warranty of any kind. IBM shall not be liable for any damages
 arising out of your or any other parties use of the sample code, even if IBM
 has been advised of the possibility of such damages. If you do not agree with
 these terms, do not use the sample code.

 Copyright IBM Corp. 2019 All Rights Reserved.

 To run, see README.md
'''

import json, os
import requests
from loggingHandler import logger
from readConfigJSON import readJSON
from ssl import SSLError
from base64 import b64encode


'''
    Function to delete all output files of one processed file with an analyzer ID through the Content Analyzer API
'''
def deleteFile(configuration_settings, analyzerId):
    # Make requests
    headers = {
        'apiKey': configuration_settings['api_key']
    }

    if "function_id" in configuration_settings:
            credentials = ('%s:%s' % (
                configuration_settings["function_id"], configuration_settings["password"]))
            encoded_credentials = b64encode(credentials.encode('ascii'))
            headers['Authorization'] = 'Basic %s' % encoded_credentials.decode("utf-8")

    get_url = configuration_settings['main_url'] + "/{0}".format(analyzerId)
    try:
        response = requests.request("DELETE", get_url, headers=headers, verify=configuration_settings['ssl_verification'])
        if response.status_code >= 400:
            logger.error("An error occurred while trying to get file status.")
            return False, response.text
        else:
            response_json = json.loads(response.text)
            return True, response_json
    except SSLError as sslerror:
        logger.error("SSL error was thrown due to certificate failure, set ssl_verification to false in configuration config.json file.")
        logger.debug(sslerror, exc_info=True)
    except Exception as ex:
        print(ex)
        logger.error("An error occured when trying to get file")
        logger.debug(ex, exc_info=True)
        pass
    return False, {}

'''
    Function to loop through each processed file and call the deleteFile function
'''
def deleteFiles():
    configuration, configuration_settings = readJSON()
    if (configuration):
        output_json_path = os.path.join(os.getcwd(), "output.json")
        if(os.path.exists(output_json_path)):
            output_json = json.load(open(output_json_path, "rb"))
            if("output_results" in output_json and len(output_json["output_results"]) > 0):
                output_json_result = output_json["output_results"]
                new_output_json_result = []
                for outresult in output_json_result:
                    result = outresult
                    try:
                        if "response" in result and "data" in json.loads(result["response"]) and "analyzerId" in \
                                json.loads(result["response"])["data"]:
                            if ("deleted" in result and result["deleted"] == False) or "deleted" not in result:
                                response = json.loads(result["response"])
                                analyzerId = response["data"]["analyzerId"]
                                status, result_response = deleteFile(configuration_settings, analyzerId)
                                if (status):
                                    result["deleted"] = True
                                else:
                                    result["deleted"] = False
                    except:
                        result["deleted"] = False
                        logger.error("No analyzerID available to delete results. The file upload may have failed. File name: {0}".format(result["filename"]))

                    # new_output_json_result.append(result)
                    output_json["output_results"] = output_json_result
                    json.dump(output_json, open("output.json", 'w'), indent=4)

                logger.info("Done deleting files on the server")
                logger.info("Delete status reported in output.json")
            else:
                logger.error("No results found that needs deleting.")
        else:
            logger.error("output.json file does not exist. No results available to delete.")


if __name__ == '__main__':
    logger.info("**************************************")
    logger.info("API Sample tool deleteFiles starting...")
    logger.info("Logs can be found in the current directory, processing.log")
    deleteFiles()

File: C#/ContentAnalyzer/DeleteFiles.cs
/*disclaimer of warranties.
 this code is sample code created by ibm corporation. ibm grants you a
 nonexclusive copyright license to use this sample code example. this
 sample code is not part of any standard ibm product and is provided to you
 solely for the purpose of assisting you in the development of your
 applications. this example has not been thoroughly tested under all
 conditions. ibm, therefore cannot guarantee nor may you imply reliability,
 serviceability, or function of these programs. the code is provided "as is",
 without warranty of any kind. ibm shall not be liable for any damages
 arising out of your or any other parties use of the sample code, even if ibm
 has been advised of the possibility of such damages. if you do not agree with
 these terms, do not use the sample code.
 copyright ibm corp. 2019 all rights reserved.
 to run, see readme.md
 * */

using System;
using System.Linq;
using System.Net;
using System.IO;
using RestSharp;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NLog;

namespace ContentAnalyzer
{
    public class DeleteFiles
    {
        private static Logger logger = LogManager.GetCurrentClassLogger();

        /// <summary>
        /// Read the configuration.json file to get the request parameters required for the delete request.
        /// If the output is downloaded already, this function will be called to delete all the related resources from data base.
        /// </summary>
        /// <param></param>
        ///<returns>"deleted":true will be saved in the same output.json file</returns>
        public bool DeleteFile()
        {

            ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;

            DirectoryInfo currentDir = new DirectoryInfo(Directory.GetCurrentDirectory());
            string configFile = currentDir.Parent.Parent.FullName + @"/NLog.config";

            string functionalID = ContentAnalyzer.UploadFiles.configParams.functionalID;
            string password = ContentAnalyzer.UploadFiles.configParams.password;
            string LDAP_userName = ContentAnalyzer.UploadFiles.configParams.LDAP_userName;
            string LDAP_password = ContentAnalyzer.UploadFiles.configParams.LDAP_password;
            string mainURL = ContentAnalyzer.UploadFiles.configParams.mainURL.Trim();
            string apiKey = ContentAnalyzer.UploadFiles.configParams.apiKey;
            string outputDirectoryPath = ContentAnalyzer.UploadFiles.configParams.outputDirectoryPath;
            string[] types = ContentAnalyzer.DownloadFiles.getTypes(ContentAnalyzer.UploadFiles.configParams.responseType);
            string auth = "";
            if (functionalID != "" && password != "")
            {
                auth = ContentAnalyzer.UploadFiles.Base64Encode(functionalID + ":" + password);
            }
            else if (LDAP_userName != "" && LDAP_password != "")
            {
                auth = ContentAnalyzer.UploadFiles.Base64Encode(LDAP_userName + ":" + LDAP_password);
            } 

            string outputDir = @"../../../output.json";
            string outputFile = File.ReadAllText(@outputDir);
            if (outputFile != "")
            {
                JObject outputFiles = JObject.Parse(@outputFile);
                JArray uploadResults = JArray.Parse(outputFiles["upload_results"].ToString());
                foreach (JObject uploadResult in uploadResults)
                {
                    JObject response = JObject.Parse(uploadResult["upload_response"].ToString());
                    JObject data = JObject.Parse(response["data"].ToString());
                    string analyzerId = data["analyzerId"].ToString();

                    Boolean allCompleted = true;
                    foreach (string type in types)
                    {
                        if (uploadResult[type].ToString() != "downloaded" && uploadResult[type].ToString() != "failed")
                        {
                            allCompleted = false;
                        }
                    }
                    if (allCompleted == true)
                    {
                        logger.Info("Deleting the output of the analyzerId: " + analyzerId);
                        string url_delete = mainURL + "/contentAnalyzer/" + analyzerId;
                        var client_delete = new RestClient(url_delete);
                        var request_delete = new RestRequest(Method.DELETE);

                        request_delete.AddHeader("Authorization", "Basic " + auth);
                        request_delete.AddHeader("apiKey", apiKey);

                        IRestResponse response_delete = client_delete.Execute(request_delete);
                        if (response_delete.StatusCode == System.Net.HttpStatusCode.OK)
                        {
                            uploadResult["deleted"] = true;
                        }
                        else
                        {
                            uploadResult["deleted"] = false;
                        }
                    }
                }
                outputFiles["upload_results"] = uploadResults;
                System.IO.File.WriteAllText(@outputDir, JsonConvert.SerializeObject(outputFiles));
            }
            return true;
        }
    }
}
